File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/VungleBannerAdapter.java
package com.ironsource.adapters.vungle;

import static com.ironsource.adapters.vungle.Constants.APP_ID;
import static com.ironsource.adapters.vungle.Constants.PLACEMENT_ID;

import android.app.Activity;
import android.text.TextUtils;
import android.view.Gravity;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ironsource.environment.ContextProvider;
import com.ironsource.mediationsdk.AdapterUtils;
import com.ironsource.mediationsdk.ISBannerSize;
import com.ironsource.mediationsdk.IronSourceBannerLayout;
import com.ironsource.mediationsdk.logger.IronLog;
import com.ironsource.mediationsdk.logger.IronSourceError;
import com.ironsource.mediationsdk.sdk.BannerSmashListener;
import com.ironsource.mediationsdk.utils.ErrorBuilder;
import com.ironsource.mediationsdk.utils.IronSourceConstants;
import com.vungle.ads.AdConfig;
import com.vungle.ads.AdSize;
import com.vungle.ads.BannerAd;
import com.vungle.ads.BannerView;
import com.vungle.ads.BaseAd;
import com.vungle.ads.BaseAdListener;
import com.vungle.ads.VungleAds;
import com.vungle.ads.VungleException;

import java.util.HashMap;
import java.util.HashSet;
import java.util.concurrent.ConcurrentHashMap;

public class VungleBannerAdapter implements AdAdapter {

    private final String providerName;

    private ConcurrentHashMap<String, BannerSmashListener> bannerSmashListeners = new ConcurrentHashMap<>();
    protected ConcurrentHashMap<String, BannerAd> bannerViews = new ConcurrentHashMap<>();

    private ISBannerSize currentBannerSize = null;
    private final Activity activity = ContextProvider.getInstance().getCurrentActiveActivity();

    public VungleBannerAdapter(@NonNull String providerName) {
        this.providerName = providerName;
    }


    public void loadBanner(String placementId,
                           IronSourceBannerLayout banner,
                           BannerSmashListener listener,
                           @Nullable String serverData) {
        IronLog.ADAPTER_API.verbose("placementId = " + placementId);

        currentBannerSize = banner.getSize();
        AdSize bannerSize = getBannerSize(currentBannerSize);

        IronLog.ADAPTER_API.verbose("bannerSize = " + bannerSize);

        if (bannerSize == null) {
            IronLog.ADAPTER_API.verbose("size not supported, size = " + banner.getSize().getDescription());
            listener.onBannerAdLoadFailed(ErrorBuilder.unsupportedBannerSize(providerName));
            return;
        }

        AdConfig adConfig = new AdConfig();
        adConfig.setAdSize(bannerSize);

        BannerAd bannerAd = new BannerAd(banner.getContext(), placementId, adConfig);
        BaseAdListener adListener = new BaseAdListener() {

            @Override
            public void error(@NonNull BaseAd baseAd, @NonNull VungleException adError) {
                IronLog.ADAPTER_CALLBACK.verbose("placementId = " + placementId + ", exception = " + adError);

                if (listener == null) {
                    IronLog.INTERNAL.verbose("listener is null");
                    return;
                }

                IronSourceError error;
                if (adError.getExceptionCode() == VungleException.NO_SERVE) {
                    error = new IronSourceError(IronSourceError.ERROR_BN_LOAD_NO_FILL, adError.getLocalizedMessage());
                } else {
                    error = ErrorBuilder.buildLoadFailedError(adError.getLocalizedMessage());
                }

                listener.onBannerAdLoadFailed(error);
            }

            @Override
            public void onAdLeftApplication(@NonNull BaseAd baseAd) {
                IronLog.ADAPTER_CALLBACK.verbose("placementId = " + placementId);

                if (listener == null) {
                    IronLog.INTERNAL.verbose("listener is null");
                    return;
                }

                listener.onBannerAdLeftApplication();
            }

            @Override
            public void adClick(@NonNull BaseAd baseAd) {
                IronLog.ADAPTER_CALLBACK.verbose("placementId = " + placementId);

                if (listener == null) {
                    IronLog.INTERNAL.verbose("listener is null");
                    return;
                }

                listener.onBannerAdClicked();
            }

            @Override
            public void adEnd(@NonNull BaseAd baseAd) {
                // no-op
            }

            @Override
            public void adImpression(@NonNull BaseAd baseAd) {
                IronLog.ADAPTER_CALLBACK.verbose("placementId = " + placementId);

                if (listener == null) {
                    IronLog.INTERNAL.verbose("listener is null");
                    return;
                }

                listener.onBannerAdShown();
            }

            @Override
            public void adStart(@NonNull BaseAd baseAd) {
                // no-op
            }

            @Override
            public void adLoaded(@NonNull BaseAd baseAd) {
                IronLog.ADAPTER_CALLBACK.verbose("placementId = " + placementId);

                if (listener == null) {
                    IronLog.INTERNAL.verbose("listener is null");
                    return;
                }

                if (currentBannerSize == null) {
                    IronLog.INTERNAL.verbose("banner size is null");
                    return;
                }

                if (!isBannerAdAvailable(placementId)) {
                    IronLog.ADAPTER_CALLBACK.error("can't play ad");
                    listener.onBannerAdLoadFailed(ErrorBuilder.buildLoadFailedError("can't play ad"));
                    return;
                }
                BannerView bannerView = bannerAd.getBannerView();

                if (bannerView != null) {
                    listener.onBannerAdLoaded(bannerAd.getBannerView(), getBannerLayoutParams(currentBannerSize));
                } else {
                    IronLog.ADAPTER_CALLBACK.error("banner view is null");
                    listener.onBannerAdLoadFailed(ErrorBuilder.buildLoadFailedError(providerName + " LoadBanner failed - banner view is null"));
                }
            }
        };
        bannerAd.setAdListener(adListener);
        bannerAd.load(serverData);

        bannerViews.put(placementId, bannerAd);
        bannerSmashListeners.put(placementId, listener);
    }

    @Override
    public void onInitializationSuccess() {
        for (BannerSmashListener listener : bannerSmashListeners.values()) {
            listener.onBannerInitSuccess();
        }
    }

    @Override
    public void onInitializationFailure(String error) {
        for (BannerSmashListener listener : bannerSmashListeners.values()) {
            listener.onBannerInitFailed(ErrorBuilder.buildInitFailedError(error, IronSourceConstants.BANNER_AD_UNIT));
        }
    }

    protected AdSize getBannerSize(ISBannerSize size) {
        switch (size.getDescription()) {
            case "BANNER":
            case "LARGE":
                return AdSize.BANNER;
            case "RECTANGLE":
                return AdSize.VUNGLE_MREC;
            case "SMART":
                return AdapterUtils.isLargeScreen(activity) ? AdSize.BANNER_LEADERBOARD : AdSize.BANNER;
        }
        return null;
    }

    protected FrameLayout.LayoutParams getBannerLayoutParams(ISBannerSize size) {
        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(0, 0);

        Activity activity = ContextProvider.getInstance().getCurrentActiveActivity();

        switch (size.getDescription()) {
            case "BANNER":
            case "LARGE":
                layoutParams = new FrameLayout.LayoutParams(AdapterUtils.dpToPixels(activity, 320), AdapterUtils.dpToPixels(activity, 50));
                break;
            case "RECTANGLE":
                layoutParams = new FrameLayout.LayoutParams(AdapterUtils.dpToPixels(activity, 300), AdapterUtils.dpToPixels(activity, 250));
                break;
            case "SMART":
                if (AdapterUtils.isLargeScreen(activity)) {
                    layoutParams = new FrameLayout.LayoutParams(AdapterUtils.dpToPixels(activity, 728), AdapterUtils.dpToPixels(activity, 90));
                } else {
                    layoutParams = new FrameLayout.LayoutParams(AdapterUtils.dpToPixels(activity, 320), AdapterUtils.dpToPixels(activity, 50));
                }

                break;
        }

        layoutParams.gravity = Gravity.CENTER;

        return layoutParams;
    }

    protected boolean isBannerAdAvailable(String placementId) {
        BannerAd ad = bannerViews.get(placementId);
        if (ad != null) {
            return VungleAds.isInitialized() && ad.canPlayAd();
        }
        return false;
    }

    public void releaseMemory() {
        for (BannerAd bannerAd : bannerViews.values()) {
            bannerAd.finishAd();
        }
        bannerViews.clear();
        bannerSmashListeners.clear();
        currentBannerSize = null;
    }

    public void initBanners(String appId, String placementId, BannerSmashListener listener) {
        if (TextUtils.isEmpty(placementId)) {
            IronLog.INTERNAL.error("Missing param - " + PLACEMENT_ID);
            listener.onBannerInitFailed(ErrorBuilder.buildInitFailedError("Missing param - " + PLACEMENT_ID, IronSourceConstants.BANNER_AD_UNIT));
            return;
        }

        if (TextUtils.isEmpty(appId)) {
            IronLog.INTERNAL.error("Missing param - " + APP_ID);
            listener.onBannerInitFailed(ErrorBuilder.buildInitFailedError("Missing param - " + APP_ID, IronSourceConstants.BANNER_AD_UNIT));
            return;
        }

        IronLog.ADAPTER_API.verbose("placementId = " + placementId + ", appId = " + appId);

        bannerSmashListeners.put(placementId, listener);
    }

    public void destroyBanner(String placementId) {
        IronLog.ADAPTER_API.verbose("Destroy Banner: placementId = " + placementId);
        BannerAd bannerAd = bannerViews.get(placementId);
        if (bannerAd != null) {
            bannerAd.finishAd();
            bannerViews.remove(placementId);
        }
        currentBannerSize = null;
    }
}

File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/banner/VungleBannerAdapter.kt
package com.ironsource.adapters.vungle.banner

import android.content.Context
import android.util.Log
import android.view.Gravity
import android.widget.FrameLayout
import com.ironsource.adapters.vungle.VungleAdapter
import com.ironsource.environment.ContextProvider
import com.ironsource.mediationsdk.AdapterUtils
import com.ironsource.mediationsdk.ISBannerSize
import com.ironsource.mediationsdk.IronSource
import com.ironsource.mediationsdk.IronSourceBannerLayout
import com.ironsource.mediationsdk.adapter.AbstractBannerAdapter
import com.ironsource.mediationsdk.logger.IronLog
import com.ironsource.mediationsdk.sdk.BannerSmashListener
import com.ironsource.mediationsdk.utils.ErrorBuilder
import com.ironsource.mediationsdk.utils.IronSourceConstants
import com.vungle.ads.BannerAd
import com.vungle.ads.BannerAdSize
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max
import kotlin.math.min

class VungleBannerAdapter(adapter: VungleAdapter) :
    AbstractBannerAdapter<VungleAdapter>(adapter) {
    private val mBannerPlacementToListenerMap: ConcurrentHashMap<String, BannerSmashListener> =
        ConcurrentHashMap()
    private val mPlacementToBannerAd: ConcurrentHashMap<String, BannerAd> =
        ConcurrentHashMap()

    //endregion

    //region Banner API
    override fun initBannerForBidding(
        appKey: String?,
        userId: String?,
        config: JSONObject,
        listener: BannerSmashListener
    ) {
        IronLog.ADAPTER_API.verbose()
        initBannersInternal(config, listener)
    }

    override fun initBanners(
        appKey: String?,
        userId: String?,
        config: JSONObject,
        listener: BannerSmashListener
    ) {
        IronLog.ADAPTER_API.verbose()
        initBannersInternal(config, listener)
    }

    private fun initBannersInternal(
        config: JSONObject,
        listener: BannerSmashListener
    ) {
        val placementId = config.optString(VungleAdapter.PLACEMENT_ID)
        val appId = config.optString(VungleAdapter.APP_ID)

        if (placementId.isNullOrEmpty()) {
            IronLog.INTERNAL.error(getAdUnitIdMissingErrorString(placementId))
            listener.onBannerInitFailed(
                ErrorBuilder.buildInitFailedError(
                    getAdUnitIdMissingErrorString(placementId),
                    IronSourceConstants.BANNER_AD_UNIT
                )
            )
            return
        }
        if (appId.isNullOrEmpty()) {
            IronLog.INTERNAL.error(getAdUnitIdMissingErrorString(appId))
            listener.onBannerInitFailed(
                ErrorBuilder.buildInitFailedError(
                    getAdUnitIdMissingErrorString(appId),
                    IronSourceConstants.BANNER_AD_UNIT
                )
            )
            return
        }
        IronLog.ADAPTER_API.verbose("placementId = $placementId, appId = $appId")

        //add to banner listener map
        mBannerPlacementToListenerMap[placementId] = listener

        when (adapter.getInitState()) {
            VungleAdapter.Companion.InitState.INIT_STATE_SUCCESS -> {
                listener.onBannerInitSuccess()
            }

            VungleAdapter.Companion.InitState.INIT_STATE_FAILED -> {
                listener.onBannerInitFailed(
                    ErrorBuilder.buildInitFailedError(
                        "Vungle SDK init failed",
                        IronSourceConstants.BANNER_AD_UNIT
                    )
                )
            }

            else -> {
                adapter.initSDK(ContextProvider.getInstance().applicationContext, appId)
            }
        }
    }

    override fun onNetworkInitCallbackSuccess() {
        mBannerPlacementToListenerMap.values.forEach { bannerListener ->
            bannerListener.onBannerInitSuccess()
        }
    }

    override fun onNetworkInitCallbackFailed(error: String?) {
        mBannerPlacementToListenerMap.values.forEach { bannerListener ->
            bannerListener.onBannerInitFailed(
                ErrorBuilder.buildInitFailedError(
                    error,
                    IronSourceConstants.BANNER_AD_UNIT
                )
            )
        }
    }

    override fun loadBannerForBidding(
        config: JSONObject,
        adData: JSONObject?,
        serverData: String?,
        banner: IronSourceBannerLayout,
        listener: BannerSmashListener
    ) {
        val placementId = config.optString(VungleAdapter.PLACEMENT_ID)
        IronLog.ADAPTER_CALLBACK.verbose("placementId = $placementId")
        loadBannerInternal(placementId, banner, listener, serverData)
    }

    override fun loadBanner(
        config: JSONObject,
        adData: JSONObject?,
        banner: IronSourceBannerLayout,
        listener: BannerSmashListener
    ) {
        val placementId = config.optString(VungleAdapter.PLACEMENT_ID)
        IronLog.ADAPTER_CALLBACK.verbose("placementId = $placementId")
        loadBannerInternal(placementId, banner, listener, null)
    }

    private fun loadBannerInternal(
        placementId: String,
        banner: IronSourceBannerLayout?,
        listener: BannerSmashListener,
        serverData: String?
    ) {
        IronLog.ADAPTER_API.verbose("placementId = $placementId")
        if (banner == null) {
            IronLog.INTERNAL.verbose("banner is null")
            listener.onBannerAdLoadFailed(ErrorBuilder.unsupportedBannerSize(adapter.providerName))
            return
        }
        val bannerSize = getBannerSize(banner.size)

        // check if banner size is null or not
        if (bannerSize == null) {
            listener.onBannerAdLoadFailed(ErrorBuilder.unsupportedBannerSize(adapter.providerName))
            return
        }
        val context = ContextProvider.getInstance().applicationContext
        val layoutParams = FrameLayout.LayoutParams(
            AdapterUtils.dpToPixels(context, bannerSize.width),
            AdapterUtils.dpToPixels(context, bannerSize.height),
            Gravity.CENTER
        )
        val vungleBannerAdListener = VungleBannerAdListener(listener, placementId, layoutParams)

        val vungleBanner = BannerAd(
            ContextProvider.getInstance().applicationContext,
            placementId,
            bannerSize
        ).apply {

            adListener = vungleBannerAdListener
        }
        mPlacementToBannerAd[placementId] = vungleBanner
        IronLog.ADAPTER_API.verbose("bannerSize = $bannerSize")
        vungleBanner.load(serverData)
    }

    override fun destroyBanner(config: JSONObject) {
        val placementId = config.optString(VungleAdapter.PLACEMENT_ID)
        IronLog.ADAPTER_API.verbose("placementId = $placementId")

        // get Vungle banner from map
        mPlacementToBannerAd[placementId]?.let { vungleBanner ->
            IronLog.ADAPTER_API.verbose(
                "destroyBanner Vungle ad, with ${VungleAdapter.PLACEMENT_ID} - $placementId"
            )

            // destroy banner
            vungleBanner.finishAd()

            // remove banner obj from the map
            mPlacementToBannerAd.remove(placementId)
        }
    }

    override fun getAdaptiveHeight(width: Int): Int {
        return getCurrentOrientationAnchoredAdaptiveBannerAdSize(
            ContextProvider.getInstance().applicationContext,
            width)?.height ?: 0
    }

    private fun getCurrentOrientationAnchoredAdaptiveBannerAdSize(
        context: Context,
        width: Int
    ): BannerAdSize? {
        return getBannerAdSize(context, width, 50, 0)
    }

    private fun getDeviceHeightInDp(context: Context, orientation: Int): Int {
        var context: Context? = context
        var orientation = orientation
        return if (context == null) {
            -1
        } else {
            if (context.applicationContext != null) {
                context = context.applicationContext
            }
            val resource = context!!.resources
            if (resource == null) {
                -1
            } else {
                val displayMetrics = resource.displayMetrics
                if (displayMetrics == null) {
                    -1
                } else {
                    val configuration = resource.configuration
                    if (configuration == null) {
                        -1
                    } else {
                        val curOrientation = configuration.orientation
                        if (orientation == 0) {
                            orientation = curOrientation
                        }
                        if (orientation == curOrientation) Math.round(
                            displayMetrics.heightPixels.toFloat() / displayMetrics.density
                        ) else Math.round(
                            displayMetrics.widthPixels.toFloat() / displayMetrics.density
                        )
                    }
                }
            }
        }
    }

    private fun getBannerAdSize(
        context: Context,
        width: Int,
        minHeight: Int,
        orientation: Int
    ): BannerAdSize? {
        var minHeight = minHeight
        var height: Int = getDeviceHeightInDp( context, orientation )
        return if (height == -1) {
            null
        } else {
            minHeight =
                min(90.0, Math.round(height.toFloat() * 0.15f).toDouble()).toInt()
            height = if (width > 655) {
                Math.round(width.toFloat() / 728.0f * 90.0f)
            } else if (width > 632) {
                81
            } else if (width > 526) {
                Math.round(width.toFloat() / 468.0f * 60.0f)
            } else if (width > 432) {
                68
            } else {
                Math.round(width.toFloat() / 320.0f * 50.0f)
            }
            height = max(min(height, minHeight), 50)
            BannerAdSize.getInlineAdaptiveBannerAdSize(width, height)
        }
    }

    private fun getBannerSize(bannerSize: ISBannerSize): BannerAdSize? {
        val vngAdSize = when (bannerSize.description) {
            "BANNER", "LARGE" -> BannerAdSize.BANNER
            "RECTANGLE" -> BannerAdSize.VUNGLE_MREC
            "SMART" ->
                (if (AdapterUtils.isLargeScreen(ContextProvider.getInstance().applicationContext)) {
                    BannerAdSize.BANNER_LEADERBOARD
                } else {
                    BannerAdSize.BANNER
                })

            else -> null
        }

        try {
            if (bannerSize.isAdaptive && vngAdSize != null) {
                val context = ContextProvider.getInstance().applicationContext
                val adaptiveSize: BannerAdSize? =
                    getCurrentOrientationAnchoredAdaptiveBannerAdSize(context, bannerSize.containerParams.width)
                Log.w("VungleBannerAdapter",
                    (((("default height - " + vngAdSize.height) +
                            " adaptive height - " + adaptiveSize?.height) +
                            " container height - " + bannerSize.containerParams.height) +
                            " default width - " + vngAdSize.width) +
                            " container width - " + bannerSize.containerParams.width
                )
                return adaptiveSize
            }
        } catch (e: Exception) {
            IronLog.INTERNAL.error("containerParams is not supported")
        }

        return vngAdSize
    }

    override fun getBannerBiddingData(
        config: JSONObject,
        adData: JSONObject?
    ): MutableMap<String?, Any?>? = adapter.getBiddingData()

    //region memory handling

    override fun releaseMemory(
        adUnit: IronSource.AD_UNIT,
        config: JSONObject?
    ) {
        IronLog.INTERNAL.verbose("adUnit = $adUnit")

        if (adUnit == IronSource.AD_UNIT.BANNER) {
            mPlacementToBannerAd.values.forEach{bannerAd ->
                bannerAd.finishAd()
            }
            mBannerPlacementToListenerMap.clear()
            mPlacementToBannerAd.clear()

        }
    }

    //endregion
}